File: 261-GraphValidTree/261-GraphValidTree.java
/*

261. Graph Valid Tree
Medium

Given n nodes labeled from 0 to n-1 and a list of undirected edges (each edge is a pair of nodes), write a function to check whether these edges make up a valid tree.

Example 1:

Input: n = 5, and edges = [[0,1], [0,2], [0,3], [1,4]]
Output: true

Example 2:

Input: n = 5, and edges = [[0,1], [1,2], [2,3], [1,3], [1,4]]
Output: false

Note: you can assume that no duplicate edges will appear in edges. Since all edges are undirected, [0,1] is the same as [1,0] and thus will not appear together in edges.

*/

class Solution {
    public boolean validTree(int n, int[][] edges) {
        // 2 pass
        // 1 - check cycle
        // 2 - check if all the nodes are connected
        if(n < 2 ) return true;
        if(edges == null || edges.length == 0) return false;
        Map<Integer, List<Integer>> graph = getGraph(n, edges);
        boolean[] vis = new boolean[n];
        if (dfs(vis, graph, 0, -1))
            return false;
        for (int i = 0; i < n; i++) 
            if (!vis[i]) return false;
        return true;
    }
    private boolean dfs(boolean[] vis, Map<Integer, List<Integer>> graph , int e, int parent){
        vis[e] = true;
        for(int n : graph.get(e)) 
            if ((vis[n] && parent != n) || (!vis[n] && dfs(vis, graph, n, e)))
                return true;
        return false;
    }
    private Map<Integer, List<Integer>> getGraph(int n, int[][] edges) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for(int i = 0; i < n; i++) graph.put(i, new ArrayList<>());
        for(int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }        
        return graph;
    }
}

File: 261-GraphValidTree/261-GraphValidTree.py
"""

261. Graph Valid Tree

You have a graph of n nodes labeled from 0 to n - 1. You are given an integer n and a list of edges where edges[i] = [ai, bi] indicates that there is an undirected edge between nodes ai and bi in the graph.

Return true if the edges of the given graph make up a valid tree, and false otherwise.

Example 1:

Input: n = 5, edges = [[0,1],[0,2],[0,3],[1,4]]
Output: true

Example 2:

Input: n = 5, edges = [[0,1],[1,2],[2,3],[1,3],[1,4]]
Output: false

Constraints:

    1 <= n <= 2000
    0 <= edges.length <= 5000
    edges[i].length == 2
    0 <= ai, bi < n
    ai != bi
    There are no self-loops or repeated edges.

"""

class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        def build_graph():
            graph = {}
            for i in range(n): graph[i]  = []
            for e in edges: 
                graph[e[0]].append(e[1])
                graph[e[1]].append(e[0])
            return graph
        graph = build_graph()
        def dfs(e, parent):
            vis[e] = True
            for n in graph.get(e):
                if (vis[n] and parent != n) or (not vis[n] and dfs(n, e)):
                    return True
            return False
        vis = [False] * n
        if dfs(0, -1): return False
        for v in vis: 
            if not v: return False
        return True
        return False
